sample_home/custom/metadata_converter.py

```diff
@@ -0,0 +1,388 @@
+""" routines to convert file metadata between SAM and MetaCat """
+import functools
+import datetime
+
+def convert_noop(x, md=None):
+    """ identity function -- no conversion """
+    return x
+
+def convert_date_mc_sam(date, md):
+    """ date format convert metacat to sam """
+    dt = datetime.datetime.fromtimestamp(date)
+    return dt.isoformat()
+
+def convert_date_sam_mc(date):
+    """ date format convert sam to metacat """
+    dt = datetime.datetime.fromisoformat(date)
+    return dt.timestamp()
+
+def convert_checksums_sam_mc(checksums):
+    """ special case: checksum conversion"""
+    res = {}
+    for cs in checksums:
+        ctype, value = cs.split(":")
+        res[ctype] = value
+    return res
+
+def convert_checksums_mc_sam(checksums, md):
+    res = []
+    for ctype in checksums:
+        res.append( "%s:%s" % (ctype, checksums[ctype]))
+    return res
+
+g_experiment = None
+
+# constants for run/subrun combination
+run_scale_factors = { 
+    "hypot": 1000000, 
+    "dune": 1000000, 
+    "hypot": 1000,
+}
+
+def convert_runs_sam_mc(runs):
+    """ special case: run/subrun  number conversion"""
+    res = []
+    m = run_scale_factors[g_experiment]
+    for r, sr, typ in runs:
+        res.append(m * r + sr)
+    return res
+
+
+def convert_runs_mc_sam( runs, md):
+    """ special case: run/subrun  number conversion"""
+    typ = md["metadata"].get("core.type", md["metadata"].get("md.type", "mc"))
+    res = []
+    m = run_scale_factors[g_experiment]
+    for r in runs:
+        rn = r // m
+        sr = r % m
+        res.append( [rn, sr, typ] )
+    return res
+
+def convert_parents_sam_mc( parents):
+    """ special case: parentage conversion"""
+    res = []
+    for i in parents:
+        res.append( {
+            "fid":  str(i["file_id"]),
+            "name": i["file_name"],
+            "namespace": g_experiment,  # XXX possible bug, what namespace?
+        })
+    return res
+
+def convert_parents_mc_sam( parents, md):
+    """ special case: parentage conversion"""
+    res = []
+    for i in parents:
+        res.append( {
+           "file_id": i["file_id"],
+           "file_name": i["name"]
+        })
+    return res
+
+
+class MetadataConverter:
+    """ converts metadata between MetaCat and Sam"""
+    def __init__(self, experiment):
+        global g_experiment
+        self.experiment = experiment
+        g_experiment = experiment
+        #
+        # conversion table:
+        #
+        self.conversion_mc_sam = {
+            "hypot": {
```

Comment:
I think this one should be mu2e, not hypot.

Reply:
So I was trying to anonymize the sample_home bits, and I globally replaced "Metacat" with "hypot"...  Yep.

Reply:
It got swapped in 2 places.  Should be better now.